### PR TITLE
Use generic math for floating point formatting

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Extracted from #98643

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs
Comment (huoyaoyuan) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: A simple `double.ToString` would fail at the following assert of this method (GitHub doesn't let me comment on it):
`typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte)`

@dotnet/jit-contrib Can you help on this for how this assert can fail? Can it be a JIT issue?

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: @huoyaoyuan are you sure it's that assert? it looks like it's actually
```cs
Debug.Assert(((precision != -1) && (precision < TNumber.MaxRoundTripDigits)) || (TNumber.FloatToBits(value) == TNumber.FloatToBits(NumberToFloat<TNumber>(ref number))));
```

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: I see in this log: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-102683-merge-12f787eac7714e67aa/Microsoft.Bcl.Numerics.Tests/1/console.8b35101e.log?helixlogtype=result

```
   at System.Number.FormatFloat[TNumber,TChar](ValueListBuilder`1& vlb, TNumber value, ReadOnlySpan`1 format, NumberFormatInfo info) in /_/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:line 537
   at System.Number.TryFormatFloat[TNumber,TChar](TNumber value, ReadOnlySpan`1 format, NumberFormatInfo info, Span`1 destination, Int32& charsWritten) in /_/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:line 517
   at System.Text.ValueStringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ReadOnlySpan`1 args) in /_/src/libraries/System.Private.CoreLib/src/System/Text/ValueStringBuilder.AppendFormat.cs:line 214
   at System.String.FormatHelper(IFormatProvider provider, String format, ReadOnlySpan`1 args) in /_/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:line 528
```

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: Does it still repro with DOTNET_ReadyToRun=0 and DOTNET_JitMinOpts=0 ? Also, what's TChar if it's not byte or char?

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: When I tried to print something locally, `TChar` was `char`. Will test with R2R/opt settings tomorrow.

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: It still reproduces when togglling R2R and MinOpts. A simple `123.456.ToString` can trigger it with corerun. I'll play with JITDump.

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: `****** DONE compiling System.Number:FormatFloat[double,ushort]`
It becomes `ushort` somewhere

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: @EgorBo It does look like a JIT issue.
The issue is in this method:
```csharp
        public static string FormatFloat<TNumber>(TNumber value, string? format, NumberFormatInfo info)
            where TNumber : unmanaged, IBinaryFloatParseAndFormatInfo<TNumber>
        {
            var vlb = new ValueListBuilder<char>(stackalloc char[CharStackBufferSize]);
            string result = FormatFloat(ref vlb, value, format, info) ?? vlb.AsSpan().ToString();
            vlb.Dispose();
            return result;
        }
```
I've confirmed with ILSpy that IL uses `char`:
```il
IL_0008: newobj instance void valuetype System.Span`1<char>::.ctor(void*, int32)
IL_001a: call valuetype System.ReadOnlySpan`1<char> System.String::op_Implicit(string)
IL_0020: call string System.Number::FormatFloat<!!TNumber, char>(valuetype System.Collections.Generic.ValueListBuilder`1<!!1>&, !!0, valuetype System.ReadOnlySpan`1<char>, class System.Globalization.NumberFormatInfo)
```

But in JITDump it says `ushort` is used:
```

STMT00008 ( ??? ... ??? )
               [000018] S-C-G------                         *  CALL      void   System.String:op_Implicit(System.String):System.ReadOnlySpan`1[ushort]
               [000024] ----------- retbuf                  +--*  LCL_ADDR  byref  V12 tmp4         [+0]
               [000017] ----------- arg1                    \--*  LCL_VAR   ref    V01 arg1         
Marked V12 as a single def temp

    [ 3]  31 (0x01f) ldarg.2
    [ 4]  32 (0x020) call 2B0003C4
In Compiler::impImportCall: opcode is call, kind=0, callRetType is ref, structSize is 0
Calling impNormStructVal on:
               [000025] -----------                         *  LCL_VAR   struct<System.ReadOnlySpan`1, 16> V12 tmp4         
resulting tree:
               [000025] -----------                         *  LCL_VAR   struct<System.ReadOnlySpan`1, 16> V12 tmp4         

lvaGrabTemp returning 13 (V13 tmp5) called for impSpillStackEnsure.


STMT00009 ( 0x01F[--] ... ??? )
               [000028] DAC-G------                         *  STORE_LCL_VAR ref    V13 tmp5         
               [000027] --C-G------                         \--*  CALL      ref    System.Number:FormatFloat[double,ushort](byref,double,System.ReadOnlySpan`1[ushort],System.Globalization.NumberFormatInfo):System.String
               [000020] ----------- arg0                       +--*  LCL_VAR   byref  V10 tmp2         
               [000022] ----------- arg1                       +--*  LCL_VAR   double V11 tmp3         
               [000025] ----------- arg2                       +--*  LCL_VAR   struct<System.ReadOnlySpan`1, 16> V12 tmp4         
               [000026] ----------- arg3                       \--*  LCL_VAR   ref    V02 arg2         
Marked V13 as a single def temp
```

Full JITDump is here: 
[JITDUMP.txt](https://github.com/dotnet/runtime/files/15457737/JITDUMP.txt)

I'm using Debug build of CLR.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: > But in JITDump it says ushort is used:

Afair, JIT renders `char` as `ushort` everywhere so it doesn't look to be the problem. Did you confirm that the assert fails because `TChar` is actually `ushort` ?

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: When I tried your branch few days ago, it failed on the other assert rather than that

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: Is JIT lowering `char` to `ushort` in some way? `Console.WriteLine(typeof(TChar).ToString())` prints `System.Char`.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: Like I said - it's how JIT renders it internally (after e.g.
```
static void Main()
{
    Foo<char>('1');
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Foo<T>(T t) { }
```
JitDump, JitDisasm and JitDisasmSummary will show the compilation as:
```
Proga:Foo[ushort](ushort)
```

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: Yes the assert does come from the line you pointed. Sorry for the noise. I'm trying to debug.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:535]: image

I've just tried locally again and it can confirm it points to that assert

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs
Comment (huoyaoyuan) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:2656]: A single character of typo for elementary school math.

**Comment by tannergooding** [comment]
Can you do some performance testing on Mono? The general process for doing so is documented here: https://github.com/dotnet/performance/blob/main/docs/benchmarking-workflow-dotnet-runtime.md

CC. @SamMonoRT

**Comment by SamMonoRT** [comment]
adding @steveisok and @vitek-karas for Mono scenario followup.

**Comment by steveisok** [comment]
@fanyang-mono can you please help with this?

**Comment by huoyaoyuan** [comment]
@EgorBot -arm64 -intel -mono

```csharp
using System.Collections.Generic;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace BenchmarkGround
{
    public class Program
    {
        public IEnumerable<object> Values => [0.0, 1.0, 1234.5678, 1.09876543211234567890e30];

        public IEnumerable<object> Formats => ["G", "F0", "F10", "E"];

        public IEnumerable<object[]> ValueAndFormats => Values.SelectMany(v => Formats.Select(f => new object[] { v, f }));

        [Benchmark]
        [ArgumentsSource(nameof(Values))]
        public string DefaultToString(double value) => value.ToString();

        [Benchmark]
        [ArgumentsSource(nameof(ValueAndFormats))]
        public string ToStringFormat(double value, string format) => value.ToString(format);
    }
}
```

**Comment by EgorBo** [comment]
@EgorBot -intel -mono --envvars MONO_ENV_OPTIONS:"--interpreter"

```csharp
using System.Collections.Generic;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace BenchmarkGround
{
    public class Program
    {
        public IEnumerable<object> Values => [0.0, 1.0, 1234.5678, 1.09876543211234567890e30];

        public IEnumerable<object> Formats => ["G", "F0", "F10", "E"];

        public IEnumerable<object[]> ValueAndFormats => Values.SelectMany(v => Formats.Select(f => new object[] { v, f }));

        [Benchmark]
        [ArgumentsSource(nameof(Values))]
        public string DefaultToString(double value) => value.ToString();

        [Benchmark]
        [ArgumentsSource(nameof(ValueAndFormats))]
        public string ToStringFormat(double value, string format) => value.ToString(format);
    }
}
```

**Comment by EgorBo** [comment]
Yeah looks like Mono-JIT and Mono interp are fine with the changes in this PR, at least for the provided benchmark. Unfortunately the bot doesn't yet support LLVM-AOT to check that as well.

**Comment by fanyang-mono** [comment]
There are instructions here to run microbenchmarks with Mono AOT. (https://github.com/dotnet/performance/blob/main/docs/benchmarking-workflow-dotnet-runtime.md#run-the-benchmarks-with-aot)

Please test it before merge. Usually, AOT takes the most hit from library changes.

**Comment by huoyaoyuan** [comment]
The instruction I used for benchmark:
```
dotnet run -c Release -f net8.0 --affinity 65535  --runtimes monoaotllvm  --aotcompilerpath "C:\Users\Meow\source\dotnet\BenchmarkTestHost\monoAOT_PR\sgen\mini\mono-sgen.exe" --customruntimepack "C:\Users\Meow\source\dotnet\BenchmarkTestHost\monoAOT_PR\pack"  --aotcompilermode llvm
```

Results:

PR:
```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3593/23H2/2023Update/SunValley3)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
.NET SDK 9.0.100-preview.4.24267.66
  [Host]     : .NET 8.0.5 (8.0.524.21615), X64 RyuJIT AVX2
  Job-FQWQJE : .NET 9.0.0 (42.42.42.42424) using MonoVM, X64 X86Base

Affinity=00000000000000001111111111111111  Runtime=MonoAOTLLVM  Toolchain=MonoAOTLLVM
```

| Method          | value                | format | Mean        | Error    | StdDev   |
|---------------- |--------------------- |------- |------------:|---------:|---------:|
| ToStringFormat  | 0                    | E      |   126.13 ns | 0.378 ns | 0.354 ns |
| ToStringFormat  | 0                    | F0     |    71.27 ns | 0.440 ns | 0.412 ns |
| ToStringFormat  | 0                    | F10    |   104.93 ns | 0.545 ns | 0.510 ns |
| ToStringFormat  | 0                    | G      |    59.68 ns | 0.524 ns | 0.490 ns |
| DefaultToString | 0                    | ?      |    54.44 ns | 0.567 ns | 0.530 ns |
| ToStringFormat  | 1                    | E      |   359.80 ns | 1.770 ns | 1.655 ns |
| ToStringFormat  | 1                    | F0     |   258.08 ns | 1.218 ns | 1.080 ns |
| ToStringFormat  | 1                    | F10    |   298.01 ns | 1.273 ns | 1.191 ns |
| ToStringFormat  | 1                    | G      |   154.68 ns | 0.510 ns | 0.477 ns |
| DefaultToString | 1                    | ?      |   148.94 ns | 0.666 ns | 0.623 ns |
| ToStringFormat  | 1234.5678            | E      |   214.97 ns | 1.181 ns | 1.105 ns |
| ToStringFormat  | 1234.5678            | F0     |   338.89 ns | 1.577 ns | 1.475 ns |
| ToStringFormat  | 1234.5678            | F10    |   501.09 ns | 2.710 ns | 2.535 ns |
| ToStringFormat  | 1234.5678            | G      |   212.78 ns | 1.048 ns | 0.929 ns |
| DefaultToString | 1234.5678            | ?      |   209.32 ns | 1.529 ns | 1.430 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | E      |   196.33 ns | 1.254 ns | 1.112 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | F0     | 1,053.69 ns | 5.062 ns | 4.487 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | F10    | 1,189.30 ns | 5.812 ns | 5.436 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | G      |   287.90 ns | 2.102 ns | 1.966 ns |
| DefaultToString | 1.098(...)6E+30 [22] | ?      |   282.25 ns | 1.599 ns | 1.496 ns |

main:

| Method          | value                | format | Mean        | Error     | StdDev    |
|---------------- |--------------------- |------- |------------:|----------:|----------:|
| ToStringFormat  | 0                    | E      |   120.53 ns |  1.899 ns |  1.776 ns |
| ToStringFormat  | 0                    | F0     |    70.56 ns |  0.576 ns |  0.481 ns |
| ToStringFormat  | 0                    | F10    |   103.54 ns |  1.347 ns |  1.260 ns |
| ToStringFormat  | 0                    | G      |    57.83 ns |  0.352 ns |  0.329 ns |
| DefaultToString | 0                    | ?      |    54.52 ns |  0.466 ns |  0.413 ns |
| ToStringFormat  | 1                    | E      |   345.41 ns |  5.116 ns |  4.786 ns |
| ToStringFormat  | 1                    | F0     |   260.28 ns |  0.619 ns |  0.549 ns |
| ToStringFormat  | 1                    | F10    |   305.23 ns |  3.058 ns |  2.860 ns |
| ToStringFormat  | 1                    | G      |   149.78 ns |  1.623 ns |  1.518 ns |
| DefaultToString | 1                    | ?      |   146.37 ns |  1.783 ns |  1.668 ns |
| ToStringFormat  | 1234.5678            | E      |   220.22 ns |  2.379 ns |  2.109 ns |
| ToStringFormat  | 1234.5678            | F0     |   351.67 ns |  2.425 ns |  2.150 ns |
| ToStringFormat  | 1234.5678            | F10    |   512.52 ns |  8.060 ns |  7.539 ns |
| ToStringFormat  | 1234.5678            | G      |   208.87 ns |  0.888 ns |  0.831 ns |
| DefaultToString | 1234.5678            | ?      |   204.73 ns |  0.748 ns |  0.699 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | E      |   194.07 ns |  2.028 ns |  1.798 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | F0     | 1,090.41 ns | 13.084 ns | 10.925 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | F10    | 1,208.15 ns | 14.228 ns | 13.309 ns |
| ToStringFormat  | 1.098(...)6E+30 [22] | G      |   290.50 ns |  2.323 ns |  2.173 ns |
| DefaultToString | 1.098(...)6E+30 [22] | ?      |   286.22 ns |  2.756 ns |  2.578 ns |

Looks no significant regression.

**Comment by tannergooding** [comment]
I can get this merged after the Mono team signs off and n the perf numbers, just want to make sure we’re not missing any important scenarios

**Comment by tannergooding** [comment]
@fanyang-mono, are there any other areas you want to see tested or does this look good for the Mono side?

**Comment by fanyang-mono** [comment]
These tests should be sufficient for mono. :)

File: src/libraries/System.Private.CoreLib/src/System/Double.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Double.cs:357]: Not going to block on this, but we should probably call this `FormatFloatingPoint` instead and it would be good to fix it in a follow up PR.

`Float` is going to be confused with the C# keyword `float`, which should be called `Single` in the BCL API surface area

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:2642]: Similar here, this should probably be `FloatingPointToBits`

**Review by tannergooding — APPROVED** [review]

**Pull request merged**
